File: RocketFuel/RocketFuel.m
/*!
 *  RocketFuel.m
 *  RocketFuel
 *
 *  Created by Ardalan Samimi on 22/10/15.
 *  Copyright © 2015 Saturn Five. All rights reserved.
 */

#import "RocketFuel.h"

@interface RocketFuel ()

@property (nonatomic) NSTask *task;
@property (nonatomic) NSString *path;
@property (nonatomic) NSArray *arguments;
@property (nonatomic, getter = shouldRelaunch) BOOL relaunch;

@end

@implementation RocketFuel

+ (instancetype)engage {
    return [[super alloc] init];
}

- (instancetype)init {
    self = [super init];

    if (self) {
        self.path = @"/usr/bin/caffeinate";
    }
    
    return self;
}

- (void)toggleSleepMode {
    if (_task) {
        [self terminate];
    } else {
        [self activate];
    }
}

- (void)activate {
    [self task];
    __weak typeof(self) weakSelf = self;
    [_task setTerminationHandler:^(NSTask * _Nonnull task) {
        [weakSelf postTerminationCleanup];
        if (weakSelf.shouldRelaunch) {
            [weakSelf activate];
            weakSelf.relaunch = NO;
        }
    }];
}

- (void)activateWithDuration:(NSInteger)duration {
    self.duration = duration;
    self.relaunch = YES;
    
    if (_task) {
        [self terminate];
        // Make sure the post termination cleanup method is called.
        // Termination handler of NSTask does not always do that.
        [self postTerminationCleanup];
    } else {
        [self activate];
    }
}

- (void)terminate {
    if ([_task isRunning]) {
        [_task terminate];
    } else {
        [self postTerminationCleanup];
    }
}

- (void)postTerminationCleanup {
    if (_task) {
        _task = nil;
        _arguments = nil;
        self.active = self.isActive;
    }
}

- (BOOL)active {
    return self.isActive;
}

- (void)setActive:(BOOL)mode {
    [self.delegate rocketFuel:self
              didChangeStatus:mode];
}

- (BOOL)isActive {
    return [_task isRunning];
}

- (NSTask *)task {
    if (!_task) {
        _task = [NSTask launchedTaskWithLaunchPath:self.path
                                         arguments:self.arguments];
        self.active = self.isActive;
    }
    
    return _task;
}

- (NSArray *)arguments {
    if (!_arguments) {
        NSString *argument;
        if (_duration) {
            argument = [NSString stringWithFormat:@"-dit %li", _duration];
        } else {
            argument = @"-di";
        }
        
        _arguments = [NSArray arrayWithObject:argument];
    }
    
    return _arguments;
}

@end

File: RocketFuel/Objects/RocketFuel.swift
//
//  RocketFuel.swift
//  RocketFuel
//
//  Created by Ardalan Samimi on 23/12/15.
//  Copyright © 2015 Ardalan Samimi. All rights reserved.
//

import Foundation

class RocketFuel: NSObject {
  
  internal var delegate: RocketFuelDelegate?
  
  internal var duration: Double = 0
  
  internal var isActive: Bool {
    return self.task?.running ?? false
  }
  
  private var task: NSTask?
  
  private var relaunch: Bool = false
  
  private var active: Bool {
    get {
      return self.isActive
    }
    set (mode) {
      self.delegate?.rocketFuel(self, didChangeStatus: mode)
    }
  }
  
  private var arguments: [String] {
    var argument: String = "-di"
    
    if self.duration > 0 {
      argument = "-dit \(self.duration)"
    }
    
    return [argument]
  }
  
  func toggle() {
    if (self.task == nil) {
      self.activate()
    } else {
      self.relaunch = false
      self.terminate()
    }
  }
  
  func activate() {
    self.task = NSTask.launchedTaskWithLaunchPath(Global.caffeinatePath, arguments: self.arguments)
    self.task?.terminationHandler = { task in
      self.didTerminate()
      // Tells RocketFuel to relaunch the task upon termination. Necessary when changing duration.
      if self.relaunch {
        self.activate()
        self.relaunch = false
      }
    }
    
    self.active = self.isActive
  }
  
  func activate(withDuration duration: Double) {
    self.duration = duration
    
    if self.task == nil {
      self.activate()
    } else {
      self.relaunch = true
      self.terminate()
      // Make sure the post termination cleanup method is called. Termination handler of NSTask does not always do that.
      self.didTerminate()
    }
    
  }
  
  func terminate() {
    if self.isActive {
      self.task?.terminate()
    } else {
      self.didTerminate()
    }
  }
  
  func didTerminate() {
    self.task = nil
    self.active = self.isActive
  }

}